File: Exercise/Very Hard/01. Cycle Detection/Code/Cycle Detection.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> graph;
vector<bool> visited;
vector<bool> rec_stack;

bool dfs(int node) {
    visited[node] = true;
    rec_stack[node] = true;

    for (int neighbor : graph[node]) {
        if (!visited[neighbor]) {
            if (dfs(neighbor)) {
                return true;
            }
        } else if (rec_stack[neighbor]) {
            return true;
        }
    }

    rec_stack[node] = false;
    return false;
}

bool has_cycle(int n, const vector<pair<int, int>>& edges) {
    graph.assign(n, vector<int>());
    visited.assign(n, false);
    rec_stack.assign(n, false);

    for (const auto& edge : edges) {
        graph[edge.first].push_back(edge.second);
    }

    for (int node = 0; node < n; ++node) {
        if (!visited[node]) {
            if (dfs(node)) {
                return true;
            }
        }
    }

    return false;
}

int main() {
    int n, m;
    cin >> n >> m;
    vector<pair<int, int>> edges(m);
    for (int i = 0; i < m; ++i) {
        cin >> edges[i].first >> edges[i].second;
    }

    if (has_cycle(n, edges)) {
        cout << "True" << endl;
    } 
    else {
        cout << "False" << endl;
    }

    return 0;
}

File: Exercise/Very Hard/01. Cycle Detection/Code/Cycle Detection.ts
function hasCycleT(n: number, edges: [number, number][]): boolean {
    const graph: number[][] = Array.from({ length: n }, () => []);
    for (const [u, v] of edges) {
        graph[u].push(v);
    }

    const visited: boolean[] = Array(n).fill(false);
    const recStack: boolean[] = Array(n).fill(false);

    function dfs(node: number): boolean {
        visited[node] = true;
        recStack[node] = true;

        for (const neighbor of graph[node]) {
            if (!visited[neighbor]) {
                if (dfs(neighbor)) {
                    return true;
                }
            } else if (recStack[neighbor]) {
                return true;
            }
        }

        recStack[node] = false;
        return false;
    }

    for (let node = 0; node < n; node++) {
        if (!visited[node]) {
            if (dfs(node)) {
                return true;
            }
        }
    }

    return false;
}

function mainT() {
    const n = 4; 
    const edges: [number, number][] = [
        [0, 1],
        [1, 2],
        [2, 3],
        [3, 1]
    ]; 

    if (hasCycleT(n, edges)) {
        console.log("True");
    } else {
        console.log("False");
    }
}

mainT();
